utils/clients/gemini/client.go

```diff
@@ -0,0 +1,279 @@
+package gemini
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/brave-intl/bat-go/settlement"
+	"github.com/brave-intl/bat-go/utils/clients"
+	"github.com/brave-intl/bat-go/utils/cryptography"
+	"github.com/brave-intl/bat-go/utils/requestutils"
+	"github.com/shengdoushi/base58"
+	"github.com/shopspring/decimal"
+)
+
+// // PrivateRequest holds all of the requisite info to complete a gemini bulk payout
+// type PrivateRequest struct {
+// 	Signature    string                   `json:"signature"` // signed over base64 payload
+// 	Payload      string                   `json:"payload"`   // base64'd
+// 	APIKey       string                   `json:"api_key"`
+// 	Transactions []settlement.Transaction `json:"transactions"`
+// }
+
+// PrivateRequestSequence handles the ability to sign a request multiple times
+type PrivateRequestSequence struct {
+	// the baseline object, corresponds to the signature in the first item
+	// must update the nonce before sending otherwise invalid signature will be encountered
+	Base       BulkPayoutPayload `json:"base"`
+	Signatures []string          `json:"signatures"` // a list of hex encoded singatures
+	APIKey     string            `json:"apikey"`     // the api key that corresponds to the checksum server side
+}
+
+// PayoutPayload contains details about transactions to be confirmed
+type PayoutPayload struct {
+	TxRef       string          `json:"tx_ref"`
+	Amount      decimal.Decimal `json:"amount"`
+	Currency    string          `json:"currency"`
+	Destination string          `json:"destination"`
+	Account     *string         `json:"account,omitempty"`
+}
+
+// AccountListPayload retrieves all accounts associated with a gemini key
+type AccountListPayload struct {
+	Request string `json:"request"`
+	Nonce   int64  `json:"nonce"`
+}
+
+// BalancesPayload retrieves all accounts associated with a gemini key
+type BalancesPayload struct {
+	Request string  `json:"request"`
+	Nonce   int64   `json:"nonce"`
+	Account *string `json:"account"`
+}
+
+// BulkPayoutPayload the payload to be base64'd
+type BulkPayoutPayload struct {
+	Request       string          `json:"request"`
+	Nonce         int64           `json:"nonce"`
+	Payouts       []PayoutPayload `json:"payouts"`
+	Account       string          `json:"account"`
+	OauthClientID string          `json:"client_id"`
+}
+
+func nonce() int64 {
```

Comment:
I don't like the name of this function, as this is potentially not a nonce when we move this to be online and multiple computes are processing lists at the same time.

Reply:
that may be, however, we're going to have to figure out a different way to present a number that always increases to gemini at that point

Reply:
:+1: